### ISSUE TITLE
短文本生成后不停止继续输出 2-3s 乱码：与字数/标点/词汇结构多因素相关（294 runs 实测）

### ISSUE THREAD

**Issue by niugtd** [studio-dots-ai/dots.tts#17]
## 现象

dots.tts-mf 在**短文本**合成时，模型说完原句后**不停止**，继续生成约 2-3 秒乱码音频——听起来像中文胡言，有时混入英文单词（如 "technology"），有时 ASR 完全识别不出（如把"我吃饭吗。"识别成"诺克萨曼蒙"）。

**关键对比**（附件 wav）：

| 输入 | 时长 | ASR 识别 | 结果 |
|---|---|---|---|
| `"我吃饭吧？"` (5字+问号) | 1.44s | "我吃饭吧。" | ✅ 干净说完即停 |
| `"我吃饭吗。"` (5字+句号) | 3.20s | **"诺克萨曼蒙"** | ⚠️ 外语乱码 |
| `"这怎么说？"` (5字+问号含"怎么") | 3.73s | (中文胡言) | ⚠️ 问号也救不了 |
| `"嗯？"` (1字+问号) | 3.20s | (无法识别) | ⚠️ 1字 100% 必乱 |

## 复现

环境：
- 模型：`rednote-hilab/dots.tts-mf`
- 硬件：DGX Spark (GB10 / aarch64 / CUDA 13.0)
- PyTorch：2.12.0+cu130
- 配置：`precision="bfloat16", optimize=True, num_steps=6, guidance_scale=1.2, speaker_scale=1.5`
- `torch.cuda.set_per_process_memory_fraction(0.75)` 已设
- `prompt_text=None`（x-vector-only cloning）
- 参考音频 ≤6s 真人录音

```python
result = runtime.generate(
    text="我吃饭吗。",  # 5字+句号 → 触发乱码
    prompt_audio_path="ref.wav",
    prompt_text=None,
    language="ZH",
    num_steps=6,
    guidance_scale=1.2,
    speaker_scale=1.5,
)
# 期望 ~1.5s，实际 3.2s，后 1.7s 是外语乱码
```

诊断脚本（带 EOS 探针）：https://github.com/niugtd/dots-tts-short-text-issue/releases/download/v0.2/test_phase5_male_refs.py

## 实测规律（294 runs 系统扫描）

经过 4 个阶段共 **294 runs** 测试，发现乱码率由**字数 + 标点 + 词汇结构**三要素综合决定。

### 规律 1：句末标点是首要因素

5 字同句干"我吃饭吧" × 3 语气词 × 3 标点（27 runs）：

| 标点 | 吧 | 吗 | 呢 |
|---|---|---|---|
| **问号 ？** | **0/3 ✅** | **0/3 ✅** | **0/3 ✅** |
| 句号 。 | 3/3 ⚠️ | 3/3 ⚠️ | 3/3 ⚠️ |
| 感叹号 ！ | 3/3 ⚠️ | 3/3 ⚠️ | 3/3 ⚠️ |

标点变种完整对照（"我吃饭吧" 5字）：
- ✅ 安全：**无标点**、**逗号中间**"我吃饭，吧。"、**问号**
- ⚠️ 危险：句号、感叹号、省略号、末尾逗号、双标点"？！"、"！？"

### 规律 2：含"怎么"的疑问句即使加问号也必乱码 ⚠️ 反例

5 字 + 问号对照（D2，24 runs）：

| 句子 | 字数+标点 | 乱码率 |
|---|---|---|
| 我吃饭吧？ | 5+？ | **0/3 ✅** |
| 你吃饭吧？ | 5+？ | **0/3 ✅** |
| 知道了吗？ | 5+？ | **0/3 ✅** |
| 饭菜好吗？ | 5+？ | **0/3 ✅** |
| 不是吗？ | 4+？ | **0/3 ✅** |
| **怎么说？** | **3+？含"怎么"** | **3/3 ⚠️** |
| **这怎么说？** | **5+？含"怎么"** | **3/3 ⚠️** |

**机制假设**：含"怎么"token 序列让 LLM 走进错误分支。可能是训练数据中此类短疑问样本稀少。

### 规律 3：1字 100% 必乱（无论标点）

| 输入 | 乱码率 |
|---|---|
| "嗯。" (1字+句号) | **3/3 ⚠️** |
| "嗯？" (1字+问号) | **3/3 ⚠️** ← 问号也救不了 |
| "对。" (1字+句号) | **3/3 ⚠️** |

### 字数 × 标点 完整规律表

| 字数 | ？结尾（不含"怎么"）| 。/！结尾 |
|---|---|---|
| 1字 | 100% 乱 | 100% 乱 |
| 2-3字 | 看词汇（"真的？"安全，"对吗？"乱）| 必乱 |
| **4字+？** | **0% 乱 ✅** | - |
| **5字+？** | **0% 乱 ✅** | - |
| 5-10字+。 | - | 17%-100% 乱 |
| **≥11字** | **0% 乱 ✅** | **0% 乱 ✅** |

### ASR 反向验证（铁证）

用 FireRedASR2S 反向转写 11 个生成样本（[asr_verification_11samples.txt](https://github.com/niugtd/dots-tts-short-text-issue/releases/download/v0.2/asr_verification_11samples.txt)），客观证明"时长异常 > 1.5x"启发式与"实际乱码"高度一致：

| 输入 | ASR 实际转写 |
|---|---|
| "我吃饭吗。" | **"诺克萨曼蒙"**（外语乱码）|
| "我吃饭吧？" | "我吃饭吧。" ✅ |
| "我吃饭呢。" | "我吃饭呢菲婷婷，我卡你卡什么？" ⚠️ |
| "我吃饭呢？" | "我吃饭呢。" ✅ |

## EOS 机制源码定位

`src/dots_tts/models/dots_tts/model.py:1586-1595`

```python
def _should_stop_after_current_audio(self, state, *, eos_threshold):
    if state.llm_hiddens is None:
        return False
    eos = (
        self.core.eos_proj(state.llm_hiddens).softmax(dim=-1)[:, -1, 1]
        > eos_threshold       # 默认 0.8
    )
    return state.end_flag or bool(eos.item())
```

通过 monkey-patch 记录每个 audio patch 的 EOS prob：

**安全句（9字"摩梭族的走婚制吗？"，13 patches）**：
```
[0,0,0,0,0,0,0,0,0,0,0, 0.0003, 0.9938]  ← 第 12 patch 突跳 0.99，干净停
```

**必乱码（5字"这是真的。"，24 patches）**：
```
[0,0,0,0,0,0.0038,...,0.0355,...,0.1188,...,0.9118]  ← 前 23 patch EOS prob 全 < 0.12
```

**灰色乱码（7字"这件事不好说。"，29 patches）**：
```
[..., 0.7708(第13), ..., 0.7705(第28), 0.9859(第29)]
       ↑ 第 13 patch 已到 0.77，但 miss 0.8 threshold
```

### 1. 更换参考音频（无效，48 runs 实测）

跨 **8 个不同参考音频**（含第三方真人录音）测试：

| ref 类别 | 数量 | 5字乱码率 | 9字乱码率 |
|---|---|---|---|
| 项目 baseline (male_ref / female_ref) | 12 runs | 100% | 0% |
| 第三方真人女声 (F0705) | 6 runs | 100% | 0% |
| **5 个第三方真人男声**（清亮青年/醇厚中年/沧桑老年/明快少年/低沉磁性中年）| **30 runs** | **100%** | **0%** |
| **跨 ref 综合** | **48 runs** | **100%** | **0%** |

**8 个 ref 乱码率完全一致 = 50%**（5字必乱 + 9字必安全）。换 ref 对乱码率零影响。

### 2. 更换 num_steps / guidance_scale / speaker_scale（无效，48 runs 实测）

| 参数 | 测试值 | 乱码率 |
|---|---|---|
| baseline | num_steps=6, guidance=1.2, speaker=1.5 | 66.7% |
| num_steps=4 | 同上但 4 | 66.7% |
| num_steps=10 | 同上但 10 | 66.7% |
| guidance=0.8 | 0.8 | 66.7% |
| guidance=2.0 | 2.0 | 50.0% |
| speaker=1.0 | 1.0 | 50.0% |
| speaker=2.0 | 2.0 | 66.7% |

**所有参数乱码率波动 50-67%（n=6 统计噪声内）**。

**机制解释**：EOS classifier 基于 LLM hidden state（`eos_proj(state.llm_hiddens).softmax(...)`），上述 4 类参数都在 LLM **下游**（flow-matching decoder / speaker embedding / 参考音频 x-vector），无法影响 EOS 判定。

### 3. 调低 eos_threshold 0.8 → 0.5（无效）

通过 monkey-patch `_should_stop_after_current_audio`：对 ≤5 字无效，因为 EOS prob 长期 < 0.12，threshold 调到 0.5 也救不了。

### 4. prompt_text 模式（看似有效但有副作用）

- 用**真实转录 + 真参考音频**（如 F0705 + "作为母亲，我没有照顾好他们的自责和愧疚。"）：5字"这是真的。"0/3 乱码，ASR 完全正确
- 用**假转录**：触发续写，CLAUDE.md 已警告过这种模式

### 5. language=None / language=YUE（看似有效但输出错误）

5字"这是真的。"0/3 乱码，但 ASR 转写为**"这是真的啊。"**——模型自由加了"啊"字，**输出错误**。

## 问题

1. **EOS classifier 训练数据是否按标点/字数分布不均？** 问号样本充足但句号短句样本稀少？含"怎么"的短疑问是不是训练数据特别少？
2. **下个模型版本会不会改善短文本稳定性？**
3. **官方有没有推荐的短文本 workaround**？目前我们的方案是脚本生成阶段 LLM 强制改写（短陈述句改问号、≥11 字符、避开"怎么"等）。但这样会改变对话原意。

## 附件

完整 release：https://github.com/niugtd/dots-tts-short-text-issue/releases/tag/v0.2

- 5 个关键 wav 样本（问号安全 / 句号乱码 / 含"怎么"乱码 / 1字必乱 / 2字+？安全）
- ASR 反向验证 txt（11 样本对照表）
- 4 阶段完整 summary.json（共 294 runs 原始数据）
- 测试脚本（含 monkey-patch EOS probe + 增量 flush）

https://github.com/niugtd/dots-tts-short-text-issue/releases/tag/v0.1 还有最初的诊断脚本和字数阈值扫描数据。

---

谢谢！如果需要更多样本、特定句式的多重复实验证、或不同硬件/模型版本的对比测试，都可以提供。

**Comment by xlians555** [comment]
感谢提供详细的信息！

你的case我复现了，大概率是因为我们训练的时候中文加language tag的比例较低，导致模型对[ZH]标签不够遵循，或者把这个tag误认为是其他语种。

不加language tag，并且没有参考文本时，在文本特别短的时候，模型难以判断是中文还是日语，我在测试中发现有一定几率会出现日语发音。

因此我认为比较好的解决办法是**language设置为None，带上prompt_text**（你提供的信息中“4. prompt_text 模式”也提示了这一点），并且需要确保prompt_text是prompt_audio的**正确转录文本**，转录文本错误时，生成内容不对是符合预期的。

希望能解决你的问题。

**Comment by Breeze-Zero** [comment]
@xlians555 Hi,
我在寻找可以解决 参考音频和要生成音频都很短（0.5～1s），比如“good”➡️“好” 的TTS方案，请问你们的模型对这类case生成稳定吗

**Comment by niugtd** [comment]
@xlians555 非常感谢！你说的方案我们实测了，确实有效。

回想起来，可能就是我们之前用法不太对——参考音频是给了，但没把它"说的什么"这个文字一起给模型，相当于只让它听声音克隆、没告诉它听到的内容。短句的情况下模型只能猜，猜错就走偏了。

**测试结果**：你推荐的 `language=None + prompt_text` 配置，"我吃饭吗。"/"这是真的。"/"这怎么说？"这 3 个 case 完全修好；但 1 字的"嗯？"还有点尾巴——3 次重复里 1 次明显乱，另外 2 次后面也带点小杂音。

后来试了下保留 `language=ZH` 不变、只加上 `prompt_text`，4 个 case 全部完全干净，包括"嗯？"。所以可能是 prompt_text 本身起了大作用，ZH tag 留着也无所谓、甚至略好一点。

总之谢谢你这条线索，帮我们绕出了之前的死胡同。

**Comment by xlians555** [comment]
> [@xlians555](https://github.com/xlians555) Hi, 我在寻找可以解决 参考音频和要生成音频都很短（0.5～1s），比如“good”➡️“好” 的TTS方案，请问你们的模型对这类case生成稳定吗

@Breeze-Zero 这种case我没有实测过，你可以在[HFSpace](https://huggingface.co/spaces/rednote-hilab/dots.tts)（soar）上先尝试一下看看能不能行
